storage/fcds/mem/mem.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2019 The Swarm Authors
+// This file is part of the Swarm library.
+//
+// The Swarm library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Swarm library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+
+package mem
+
+import (
+	"sync"
+
+	"github.com/ethersphere/swarm/chunk"
+	"github.com/ethersphere/swarm/storage/fcds"
+)
+
+var _ fcds.MetaStore = new(MetaStore)
+
+// MetaStore is the simplest in-memory implementation of FCDS MetaStore.
+// It is meant to be used as the reference implementation.
+type MetaStore struct {
+	meta map[string]*fcds.Meta
+	free map[uint8]map[int64]struct{}
+	mu   sync.RWMutex
+}
+
+// NewMetaStore constructs a new MetaStore.
+func NewMetaStore() (s *MetaStore) {
+	free := make(map[uint8]map[int64]struct{})
+	for shard := uint8(0); shard < 255; shard++ {
+		free[shard] = make(map[int64]struct{})
+	}
+	return &MetaStore{
+		meta: make(map[string]*fcds.Meta),
+		free: free,
+	}
+}
+
+// Get returns chunk meta information.
+func (s *MetaStore) Get(addr chunk.Address) (m *fcds.Meta, err error) {
+	s.mu.RLock()
+	m = s.meta[string(addr)]
+	s.mu.RUnlock()
+	if m == nil {
+		return nil, chunk.ErrChunkNotFound
+	}
+	return m, nil
+}
+
+// Set adds a new chunk meta information for a shard.
+// Reclaimed flag denotes that the chunk is at the place of
+// already deleted chunk, not appended to the end of the file.
+func (s *MetaStore) Set(addr chunk.Address, shard uint8, reclaimed bool, m *fcds.Meta) (err error) {
+	s.mu.Lock()
```

Comment:
defer unlock?

Reply:
Defer comes with a cost, and I wanted to have as good as possible performance for measurements.